src/commonMain/kotlin/app/cash/turbine/flow.kt

```diff
@@ -0,0 +1,174 @@
+ */
+@Deprecated("Timeout parameter removed. Use runTest which has a timeout or wrap in withTimeout.",
+  ReplaceWith("this.test(validate)"),
+  DeprecationLevel.ERROR,
+)
+@Suppress("UNUSED_PARAMETER")
+public suspend fun <T> Flow<T>.test(
+  timeout: Duration = 1.seconds,
+  validate: suspend ReceiveTurbine<T>.() -> Unit
+) {
+  test(validate)
+}
+
+/**
+ * Terminal flow operator that collects events from given flow and allows the [validate] lambda to
+ * consume and assert properties on them in order. If any exception occurs during validation the
+ * exception is rethrown from this method.
+ *
+ * ```kotlin
+ * flowOf("one", "two").test {
+ *   assertEquals("one", expectItem())
+ *   assertEquals("two", expectItem())
+ *   expectComplete()
+ * }
+ * ```
+ */
+public suspend fun <T> Flow<T>.test(
+  validate: suspend ReceiveTurbine<T>.() -> Unit
+) {
+  coroutineScope {
+    collectTurbineIn(this).apply {
+      validate()
+      cancel()
+      ensureAllEventsConsumed()
+    }
+  }
+}
+
+/**
+ * Terminal flow operator that collects events from given flow and returns a [ReceiveTurbine] for
+ * consuming and asserting properties on them in order. If any exception occurs during validation the
+ * exception is rethrown from this method.
+ *
+ * ```kotlin
+ * val turbine = flowOf("one", "two").testIn(this)
+ * assertEquals("one", turbine.expectItem())
+ * assertEquals("two", turbine.expectItem())
+ * turbine.expectComplete()
+ * ```
+ *
+ * Unlike [test] which automatically cancels the flow at the end of the lambda, the returned
+ * [ReceiveTurbine] must either consume a terminal event (complete or error) or be explicitly canceled.
+ */
+public fun <T> Flow<T>.testIn(scope: CoroutineScope): ReceiveTurbine<T> {
+  val turbine = collectTurbineIn(scope)
+
+  scope.coroutineContext.job.invokeOnCompletion { exception ->
+    if (debug) println("Scope ending ${exception ?: ""}")
+
+    // Only validate events were consumed if the scope is exiting normally.
+    if (exception == null) {
+      turbine.ensureAllEventsConsumed()
+    }
+  }
+
+  return turbine
+}
+
+private fun <T> Flow<T>.collectTurbineIn(scope: CoroutineScope): Turbine<T> {
+  val (channel, job) = scope.jobify { collectIntoChannel(this) }
+
+  return TurbineImpl(channel, job)
+}
+
+internal fun <T> Flow<T>.collectIntoChannel(scope: CoroutineScope): Channel<T> {
+  val output = Channel<T>(Channel.UNLIMITED)
+  val job = scope.launch(start = CoroutineStart.UNDISPATCHED) {
+    try {
+      collect { output.trySend(it) }
+      output.close()
+    } catch (e: Exception) {
+      output.close(e)
+    }
+  }
+
+  return object : Channel<T> by output {
+    override fun cancel(cause: CancellationException?) {
+      job.cancel()
+      output.close(cause)
+    }
+
+    override fun close(cause: Throwable?): Boolean {
+      job.cancel()
+      return output.close(cause)
+    }
+  }
+}
+
+private inline fun <T> CoroutineScope.jobify(crossinline block: CoroutineScope.() -> T) : Pair<T, Job> {
```

Comment:
Seems to be only used once. Inline to call site?

Reply:
I just wish something like this were in the stdlib 😢